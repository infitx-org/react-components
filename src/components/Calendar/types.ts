export type PossibleDate = Date | undefined;

export type PossibleDay = string | undefined;

export type Matrix = PossibleDay[][];

export type DisabledDays = (day: Date) => boolean;

export type DateRange = [PossibleDate, PossibleDate];
