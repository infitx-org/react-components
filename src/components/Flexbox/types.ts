export enum TopBottom {
  "top" = "flex-start",
  "bottom" = "flex-end",
}
export enum LeftRight {
  "left" = "flex-start",
  "right" = "flex-end",
}

export enum Mappers {
  "top" = "flex-start",
  "bottom" = "flex-end",
  "left" = "flex-start",
  "right" = "flex-end",
}

export type AlignItems =
  | "flex-start"
  | "flex-end"
  | "center"
  | "baseline"
  | "stretch";

export type JustifyContent =
  | "flex-start"
  | "flex-end"
  | "center"
  | "space-between"
  | "space-around"
  | "space-evenly";

export type AlignWithMap<T extends string> = `${AlignItems}` | `${T}`;
export type JustifyWithMap<T extends string> = `${JustifyContent}` | `${T}`;
