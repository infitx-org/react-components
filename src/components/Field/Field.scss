@import "../scss/vars/colors.scss";
@import "../scss/vars/border.scss";
@import "../scss/vars/background.scss";
@import "../scss/vars/shadow.scss";
@import "../scss/vars/general.scss";
@import "../scss/base.scss";

.rc-field {
  @include rc-component;
  border-radius: $border-radius;
  position: relative;
  display: inline-flex;
  align-items: center;
  width: 200px;
}
.rc-field--open {
  box-shadow: $shadow-size;
  padding: 0px $component-open-size;
}
.rc-field--disabled {
  opacity: $component-opacity--disabled;
  pointer-events: none;
  cursor: auto;
}

.rc-field--small {
  height: 31px;
  padding-left: 10px;
  padding-right: 6px;
  font-size: 12px;
}
.rc-field--medium {
  height: 38px;
  padding-left: 13px;
  padding-right: 8px;
  font-size: 13px;
}
.rc-field--large {
  height: 45px;
  padding-left: 16px;
  padding-right: 10px;
  font-size: 14px;
}

.rc-field__shadow {
  color: transparent;
}
.rc-field--open {
  &.rc-field__shadow {
    color: $shadow-color--default;
  }
  &.rc-field__shadow--pending {
    color: $shadow-color--pending;
  }
  &.rc-field__shadow--required {
    color: $shadow-color--required;
  }
  &.rc-field__shadow--invalid {
    color: $shadow-color--invalid;
  }
}

.rc-field__background {
  background: $background--default;
}
.rc-field__background--open {
  background: $background--open;
}
.rc-field__background--pending {
  background: $background--pending;
}
.rc-field__background--required {
  background: $background--required;
}
.rc-field__background--invalid {
  background: $background--invalid;
}

.rc-field__borders {
  border-left: 2px solid;
  border-right: 2px solid;
  border-bottom-color: $border-color--default;
  border-right-width: 1px;
  border-left-width: 1px;
  border-top-width: 1px;
  border-bottom-width: 2px;
  border-bottom-style: solid;
  border-top-style: solid;
  border-right-color: #e5e5e5;
  border-left-color: #e5e5e5;
  border-top-color: #e5e5e5;
}
.rc-field__borders--open {
}
.rc-field__borders--pending {
  border-bottom-color: $border-color--pending;
}
.rc-field__borders--required {
  border-bottom-color: $border-color--required;
}
.rc-field__borders--invalid {
  border-bottom-color: $border-color--invalid;
}
