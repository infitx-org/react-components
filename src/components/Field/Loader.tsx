import Spinner from "../Spinner";
import { Size } from "../types";
import { getIconSizeByComponentSize } from "../shared";

interface LoaderProps {
  size: `${Size}`;
}

export default function Loader({ size = Size.Large }: LoaderProps) {
  return (
    <div className="mb-input__inner-icon mb-loader">
      <Spinner size={getIconSizeByComponentSize(size)} />
    </div>
  );
}
