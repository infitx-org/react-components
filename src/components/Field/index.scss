@use "./Field.scss" as *;
@use "./components/Loader.scss" as *;
@use "./components/Placeholder.scss" as *;
@use "./components/InvalidIcon.scss" as *;
