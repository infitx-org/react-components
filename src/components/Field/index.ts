export { default, FieldProps } from "./Field";
export { default as Loader, LoaderProps } from "./components/Loader";
export {
  default as Placeholder,
  PlaceholderProps,
} from "./components/Placeholder";
