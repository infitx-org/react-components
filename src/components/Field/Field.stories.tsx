import React from "react";
import Field from ".";

export default {
  title: "Field",
};

export const DefaultField = () => {
  return (
    <Field>
      <div
        style={{
          height: "14px",
          width: "14px",
          borderRadius: "50%",
          background: "#f33",
        }}
      />
      <span style={{ marginLeft: "10px", color: "#F33" }}>test</span>
    </Field>
  );
};

const CreditCardExampleField = ({ value, onChange }) => {
  function formatValue(unformatted: string): string {
    const chunks = unformatted.replace(/ /g, "").match(/.{1,4}/g) || [];
    return chunks.join(" ");
  }
  return (
    <Field pending required>
      <input
        type="text"
        inputMode="numeric"
        value={formatValue(value)}
        onChange={onChange}
        style={{ appearance: "none", border: "none", outline: "none" }}
      />
    </Field>
  );
};

export const DefaultSimpleMultiField = () => {
  const [value, setValue] = React.useState("1234123412341234");
  return (
    <CreditCardExampleField
      value={value}
      onChange={(e) => setValue(e.target.value)}
    />
  );
};

const Template = (args) => (
  <Field {...args}>
    <span>children render here</span>
  </Field>
);

export const Default = Template.bind({});
Default.args = {
  size: "large",
  className: undefined,
  required: false,
  invalid: false,
  pending: false,
};

export const Disabled = Template.bind({});
Disabled.args = {
  disabled: true,
};

export const Pending = Template.bind({});
Pending.args = {
  pending: true,
};

export const Required = Template.bind({});
Required.args = {
  required: true,
};

export const Invalid = Template.bind({});
Invalid.args = {
  invalid: true,
};

export const Small = Template.bind({});
Small.args = {
  size: "small",
};

export const Medium = Template.bind({});
Medium.args = {
  size: "medium",
};

export const Large = Template.bind({});
Large.args = {
  size: "large",
};
