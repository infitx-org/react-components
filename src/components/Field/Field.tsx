import React from "react";
import classnames from "classnames";
import { Kind, InputSize } from "types";
import useOnClickOutside from "hooks/useOnClickOutside";
import "./Field.scss";

export type FieldProps = {
  kind?: `${Kind}`;
  size?: `${InputSize}`;
  label?: string;
  required?: boolean;
  pending?: boolean;
  invalid?: boolean;
  disabled?: boolean;
  children: React.ReactNode;
  focused?: boolean;
  className?: string;
  onClick?: (e: React.MouseEvent<HTMLDivElement>) => void;
  onClickOutside?: (e: MouseEvent) => void;
};

function Field({
  kind = Kind.Primary,
  size = InputSize.Large,
  label,
  disabled,
  required,
  pending,
  invalid,
  children,
  focused,
  className,
  onClick,
  onClickOutside,
}: FieldProps): JSX.Element {
  const fieldRef = React.useRef<HTMLDivElement>(null);

  useOnClickOutside(fieldRef, onClickOutside);

  const fieldClassname = classnames([
    "rc-field",
    kind && `rc-field--${kind}`,
    "rc-field__borders",
    "rc-field__background",
    "rc-field__shadow",
    size === InputSize.Small && "rc-field--small",
    size === InputSize.Medium && "rc-field--medium",
    size === InputSize.Large && "rc-field--large",
    focused &&
      "rc-field--open rc-field__borders--open rc-field__background--open rc-field__shadow--open",
    disabled &&
      "rc-field--disabled rc-field__borders--disabled rc-field__background--disabled",
    pending &&
      "rc-field--pending rc-field__borders--pending rc-field__background--pending rc-field__shadow--pending",
    invalid &&
      "rc-field--invalid rc-field__borders--invalid rc-field__background--invalid rc-field__shadow--invalid",
    required &&
      "rc-field--required rc-field__borders--required rc-field__background--required rc-field__shadow--required",
    className,
  ]);

  return (
    <div>
      {label && <label>{label}</label>}
      <div
        className={fieldClassname}
        onClick={onClick}
        role="presentation"
        ref={fieldRef}
      >
        {children}
      </div>
    </div>
  );
}

export default React.memo(Field);
