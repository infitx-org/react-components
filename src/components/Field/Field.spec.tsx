import { render } from "@testing-library/react";
import "@testing-library/jest-dom/extend-expect";
import userEvent from "@testing-library/user-event";
import { InputSize, Kind } from "types";
import Field from "./Field";

describe("tests the Field props", () => {
  it("renders the Field", () => {
    const { container } = render(<Field>test</Field>);
    expect(container.querySelector(".rc-field")).toBeTruthy();
  });

  it("renders the children", () => {
    const { container } = render(
      <Field>
        <span>Child</span>
      </Field>
    );
    expect(container.querySelector("span")).toHaveTextContent("Child");
  });

  it("renders the classname", () => {
    const { container } = render(<Field className="test">test</Field>);
    expect(container.querySelector(".test")).toBeTruthy();
  });

  it("renders as disabled", () => {
    const { container } = render(<Field disabled>test</Field>);
    expect(container.querySelector(".rc-field--disabled")).toBeTruthy();
  });

  it("renders as required", () => {
    const { container } = render(<Field required>test</Field>);
    expect(container.querySelector(".rc-field--required")).toBeTruthy();
  });

  it("renders as invalid", () => {
    const { container } = render(<Field invalid>test</Field>);
    expect(container.querySelector(".rc-field--invalid")).toBeTruthy();
  });

  it("renders as focused", () => {
    const { container } = render(<Field focused>test</Field>);
    expect(container.querySelector(".rc-field--focused")).toBeTruthy();
  });

  it("renders the small, medium, large sizes", () => {
    Object.values(InputSize).forEach((size) => {
      const { container } = render(<Field size={size}>test</Field>);
      expect(container.querySelector(`.rc-field--${size}`)).toBeTruthy();
    });
  });

  it("renders the kinds", () => {
    Object.values(Kind).forEach((kind) => {
      const { container } = render(<Field kind={kind}>test</Field>);
      expect(container.querySelector(`.rc-field--${kind}`)).toBeTruthy();
    });
  });

  it("triggers onClick when clicking", () => {
    const mockFn = jest.fn();
    const { container } = render(<Field onClick={mockFn}>test</Field>);
    userEvent.click(container.querySelector(".rc-field") as Element);
    expect(mockFn).toHaveBeenCalled();
  });
});

// Snapshot testing
it("renders the Field correctly when multiple props are set", () => {
  const { container } = render(
    <Field required invalid disabled>
      test
    </Field>
  );
  expect(container).toMatchSnapshot();
});
