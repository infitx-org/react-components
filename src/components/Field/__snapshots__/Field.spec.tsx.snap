// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the Field correctly when multiple props are set 1`] = `
<div>
  <div
    class="rc-field rc-field--primary rc-field--large rc-field--disabled rc-field--pending rc-field--invalid rc-field--required"
    role="presentation"
  >
    test
  </div>
</div>
`;
